src/libraries/Common/src/System/Number.Formatting.Common.cs

```diff
@@ -0,0 +1,1185 @@
+                    // Fast path for sole symbol, e.g. "D"
+                    if (format.Length == 1)
+                    {
+                        digits = -1;
+                        return c;
+                    }
+
+                    if (format.Length == 2)
+                    {
+                        // Fast path for symbol and single digit, e.g. "X4"
+                        int d = format[1] - '0';
+                        if ((uint)d < 10)
+                        {
+                            digits = d;
+                            return c;
+                        }
+                    }
+                    else if (format.Length == 3)
+                    {
+                        // Fast path for symbol and double digit, e.g. "F12"
+                        int d1 = format[1] - '0', d2 = format[2] - '0';
+                        if ((uint)d1 < 10 && (uint)d2 < 10)
+                        {
+                            digits = d1 * 10 + d2;
+                            return c;
+                        }
+                    }
+
+                    // Fallback for symbol and any length digits.  The digits value must be >= 0 && <= 999_999_999,
+                    // but it can begin with any number of 0s, and thus we may need to check more than 9
+                    // digits.  Further, for compat, we need to stop when we hit a null char.
+                    int n = 0;
+                    int i = 1;
+                    while ((uint)i < (uint)format.Length && char.IsAsciiDigit(format[i]))
+                    {
+                        // Check if we are about to overflow past our limit of 9 digits
+                        if (n >= 100_000_000)
+                        {
+                            ThrowHelper.ThrowFormatException_BadFormatSpecifier();
+                        }
+                        n = (n * 10) + format[i++] - '0';
+                    }
+
+                    // If we're at the end of the digits rather than having stopped because we hit something
+                    // other than a digit or overflowed, return the standard format info.
+                    if ((uint)i >= (uint)format.Length || format[i] == '\0')
+                    {
+                        digits = n;
+                        return c;
+                    }
+                }
+            }
+
+            // Default empty format to be "G"; custom format is signified with '\0'.
+            digits = -1;
+            return format.Length == 0 || c == '\0' ? // For compat, treat '\0' as the end of the specifier, even if the specifier extends beyond it.
+                'G' :
+                '\0';
+        }
+
+        internal static unsafe void NumberToString<TChar>(ref ValueListBuilder<TChar> vlb, ref NumberBuffer number, char format, int nMaxDigits, NumberFormatInfo info) where TChar : unmanaged, IUtfChar<TChar>
+        {
+            Debug.Assert(typeof(TChar) == typeof(char) || typeof(TChar) == typeof(byte));
```

Comment:
This assert is painful for polyfill. How about switching to `sizeof`?
`IBinaryInteger` provides similar performance with `IUtfChar` on CLR, but on mono it may cause regression.

Reply:
How is it painful, I'm not understanding?

Reply:
S.R.Numerics can't call these methods with byte or char directly, but a polyfill type that implements an internal definition of IUtfChar instead. I have done this in previous PR. There's no such type check in parsing logic.